### ISSUE TITLE
Transforms not honoring destination paths

### ISSUE THREAD

**Issue by orbitbot** [futurechan/gulp-asset-transform#12]
Hi! Nice project, I ran in to a similar issue as you did in a comment somewhere here on Github with LESS compilation using Usemin, and this seems to solve that problem nicely.

I have however run into another fundamental issue: This library is not honoring the destination paths  defined in the ```<!-- at:id >> path/to/somewhere -->``` comments in html files. I have made a small gist to demonstrate the issue, which should be correct as far as I understand the documentation and examples:

https://gist.github.com/orbitbot/772207c04ba57b82fb75

Based on the documentation, I would expect that a definition as follows (from the gist index.html):
```
    <!-- at:css >> css/normalize.css -->
    <link rel="stylesheet" href="normalize.css">
    <!-- at:end -->
```
... should output a file to ```dist/css/normalize.css``` according to my gulpfile configuration. However, it seems that the path is lost somehow, and all my compiled files end up directly under ```dist/```, in this case ```dist/normalize.css``` instead of ```dist/css/normalize.css```.

**Comment by futurechan** [comment]
Thanks for the interest in this project. This does seem like a fundamental issue. What version of the project are you using and what is your gulp configuration?

**Comment by futurechan** [comment]
Well, I changed the comment directive to
```
<!-- at:css >> css/blah.css -->
    <link rel="stylesheet" href="normalize.css">
<!-- at:end -->
```

and normalize.css was still output to the dist folder. It looks like I released a bug. I'll have an update very soon. Sorry.

**Comment by orbitbot** [comment]
The package.json is included, if you're looking through the browser it ended up all the way below the normalize.css file. The issue should be reproduceable with 
```
git clone
npm install
gulp
```

**Comment by orbitbot** [comment]
Thanks for the quick reply! I'm not sure if this is related or actually intended functionality, and I can file another issue if not, but with a config like the following:

```
      css: {
        stream: function(filestream, outputFilename){
            return filestream
                .pipe(plugins.size({ title: 'css', showFiles: true }))
                .pipe(gulp.dest('dist/'))
                .pipe(minifyCss())
                .pipe(plugins.rename({ suffix: '.min' }))
                .pipe(plugins.size({ title: 'minfied css', showFiles: true}))
                .pipe(gulp.dest('dist/'));
        }
      },
```
... the ```index.html``` file's header will refer to the minified filename like so: ```<link rel="stylesheet" type="text/css" href="normalize.min.css">```, instead of the 'normal' file specified in the source .html file:
```
    <!-- at:css >> normalize.css -->
    <link rel="stylesheet" href="css/normalize.css">
    <!-- at:end -->
```
Otherwise, might this somehow be related to me not terminating the stream with the ```outputfile``` parameter somehow? Without having a look at the code in detail this feels a bit odd.

**Comment by futurechan** [comment]
Yes, this does look like a different issues.

To be clear, my temporary workaround for your first issue is

```
gulp.task('default', function() {
  return gulp.src(paths.index)
    .pipe(at({
      css: {
        tasks: [size({ title: 'css', showFiles: true }), 'concat']
      }
    }))
    .pipe(gulp.dest('dist/'));
});

**Comment by futurechan** [comment]
Regarding your second issue. AT has a "default" tag generation based on assumptions and the output of the task stream. However, you can override that with the custom tag templating.

**Comment by orbitbot** [comment]
I guess I didn't fully grasp the consequences of the implicit tag reference templates that were mentioned in the documentation when I read through them. I have no issue with using the minified file as such, it's mainly that I ended up with a different result to what I thought I was configuring and didn't understand the reasons for it.

Thanks for the workaround tip, I'll give it a try!

**Comment by futurechan** [comment]
Just to clarify, when using implicit tags, the name comes from the comment directive and the extension used within that directive. ```<!-- at:css >> css/normalize.css -->``` uses the default template for a ```.css``` file and would normally use the name ```css/normalize.css```. However, some plugins like gulp-rev need to rename the file. In this case, using ```css/normalize.css``` doesn't really make sense. AT will honor the file name from ```File.path```. 

In your case, if you want the output stream to honor the original name, try removing ``` .pipe(plugins.rename({ suffix: '.min' }))```. 

By the way, did the workaround work for you?

**Comment by orbitbot** [comment]
1) The workaround with the tasks setup works using ```'concat'```. 

I'm setting up a toolchain for making modifications to a base html file with CSS and LESS with gulp. I guess I was in the special case of not combining multiple CSS files into one, which is why I left the ```concat``` task out, but it's probably good practice to have there to make the CSS tasks completely generic.

2) Removing only ```.pipe(plugins.rename({ suffix: '.min' }))``` did not help with the streams approach, I still end up with files in the ```dist/``` folder. However, similar to the ```'concat'``` workaround, using ```.pipe(plugins.concat(outputFilename));``` as in the example will put the compiled files in the correct folders.

I guess I can work around my problem with the ```concat```-related tasks in place, and separate build tasks for minification and renaming.

**Comment by futurechan** [comment]
Yeah, it looks like I expect concat to be somewhere in the pipeline. I'm not sure what to do just yet and I'm open to suggestions. One thought is - in the case of "single file" pipelines, implement gulp-rename. I could also allow the user to specify pipelines without an output filename or with only an output directory.

**Comment by orbitbot** [comment]
My original assumption was that the filename specified in ```<!-- at:css >> css/normalize.css -->``` in the template html would simply be processed into the resulting html file without any kind of renaming support. Actually supporting renaming on the fly is a neat feature, the original issue for me was that the folders got lost with the filename transformation and without the concat. I think that assuming that the template html's directories are there for a reason is a sensible default and would allow my original approach (ie. prepend to any output filename?).

 - In the case of "single file" pipelines, I could implement gulp-rename
I guess this is dependent on how the processing is done currently, do you already somehow keep track of how many source files are in a directive before starting processing?

 - I could also allow the user to specify pipelines without an output filename or with only an output directory.
This would still force users to specify a filename somehow, though, wouldn't that be problematic whenever you have multiple directives of the same type? Keeping the filename specification in the html directive is much cleaner than having to specify a separate tag_template_name and task for each filename...

 - I could also forcibly append concat to the beginning of any task.
Doesn't this make too many assumptions about what could be included in a html template?

**Comment by orbitbot** [comment]
Mea culpa, pressed wrong button when commenting...

**Comment by futurechan** [comment]
+ Yes, I know all assets within a directive before processing begins
+ If the user didn't specify an output filename, and if concat is not involved, then each asset's current filename/relative path could be reused. If the user specifies an output directory, we could use that.
+ I agree, which is why I haven't added 'concat' to date. Unfortunately, if multiple assets are listed within a directive and 'concat' is not used, it's a bit ambiguous what is intended. I'd really like to sure that up. 

Thanks for the feedback btw.

**Comment by orbitbot** [comment]
> If the user didn't specify an output filename, and if concat is not involved, then each asset's current filename/relative path could be reused. If the user specifies an output directory, we could use that.

I think this would be very useful and a good flexible solution.

> I agree, which is why I haven't added 'concat' to date. Unfortunately, if multiple assets are listed within a directive and 'concat' is not used, it's a bit ambiguous what is intended. I'd really like to sure that up.

In line with the above implementation suggestion, maybe one solution would be to apply the task transformations to each file individually by default, respecting either relative path or a directory path in the directive, and only respect the output filename in the directive in case concat is one of the tasks? That would leave the configuration up to the end users instead of having to guess what has been intended.

> Thanks for the feedback btw.

NP! Glad to help.

**Comment by futurechan** [comment]
The files are processed individually until concat. Basically, I parse the file names within a directive, create a glob stream from them, and pass that to the first task, which is piped to the second until completion.

As it stands today, it's hard to know if concat will or will not be used and if a rename is required.

**Comment by orbitbot** [comment]
Would it be reasonable to have it work like the following?

Single file case:
```
1) <!-- at:css >>  -->
    => if no output destination specified, use relative path to input html
2) <!-- at:css >> my/new/folder/ -->
    => if a folder or path with unknown or unrecognized file extension is specified,
       use that output folder instead of relative to input
3) <!-- at:css >> my/new/path/to/normalize.css -->
    => if a full path is specified, rename file to new name
```
... and similarly for multiple file input:
```
1) & 2) identical.
3) <!-- at:css >> my/new/path/to/normalize.css -->
    => if a full path is specified, assume that user will call concat
       with that path as input, otherwise identical to 2)
```
Do you feel that user intent would still be ambiguous with something like this?

**Comment by futurechan** [comment]
I feel like this safely builds upon what is already provided, but I have one change - let's drop the ```>>``` from ```1)```.
